compiler/vm/vm.nim

```diff
@@ -71,34 +74,105 @@ when hasFFI:
 const
   errIllegalConvFromXtoY = "illegal conversion from '$1' to '$2'"
 
+type AddressCheckFailReason = enum
+  acfrOutOfBounds ## Address points to outside the vm's memory
+  acfrTypeMismatch ## It's not legal to write to the target location with handle of given type
+
+func checkValid(mm: VmMemoryManager, a: MemoryHandle, typ: PVmType, r: var AddressCheckFailReason): bool =
+  ## Test if a typed handle with address `a` and type `typ` would be valid
+  let rp = cast[int](a.toPointer)
+
+  r = acfrOutOfBounds
+
+  var found = false
+  for x in mm.regions.items:
+    if rp in x.start..(x.start+x.len):
+      if x.typ == typ:
+        found = true
+      else:
+        r = acfrTypeMismatch
+      break
+
+  return found
+
+
+func getFieldHandle(h: TypedHandle, idx: FieldPosition): TypedHandle {.inline.} =
+  let f = h.typ.objFields[idx]
+  h.getSubHandle(f.offset, f.typ)
+
+
+# XXX: I don't think exception unmarshaling/marhshaling is really needed. And even if it is,
+# use `unmarshalGeneric` instead
+func unmarshalException(e: TypedHandle, nimType: PType): PNode =
```

Comment:
A small tip, if you are planning to replace something, attach `{.deprecated.}` to it so you get all the call sites reported, which should simplify your work.